### PR TITLE
feat(cketh/ckerc20): Deposit events with subaccounts

### PR DESCRIPTION
**Pull request by gregorydemay**
Parse deposit events originating from the smart contract introduced in #2143, which additionally contain an optional ledger subaccount. The events from that smart contract are not yet scraped, which will be done in a follow-up PR.

### CODE REVIEW

File: rs/ethereum/cketh/minter/src/eth_logs/mod.rs [hunk 1]
```diff
@@ -378,6 +441,30 @@ impl TryFrom<LogEntry> for ReceivedEvent {
...
+            )),
+        });
+    }
+    let mut result = Vec::with_capacity(N);
+    for chunk in data.chunks_exact(32) {
+        let mut word = [0; 32];
+        word.copy_from_slice(chunk);
+        result.push(word);
+    }
+    Ok(result.try_into().unwrap())
```
Comment (THLO) [inline rs/ethereum/cketh/minter/src/eth_logs/mod.rs:465]: Just for my understanding, if you call `unwrap` anyway, why do you need `try_into`? Is it because `into` doesn't exist?

File: rs/ethereum/cketh/minter/src/eth_logs/mod.rs [hunk 2]
```diff
@@ -378,6 +441,30 @@ impl TryFrom<LogEntry> for ReceivedEvent {
     }
 }
 
+fn parse_data_into_32_bytes_words<const N: usize>(
```
Comment (THLO) [inline rs/ethereum/cketh/minter/src/eth_logs/mod.rs:444]: Nit: Since it's 32-**byte** words, how about calling the function `parse_data_into_32_byte_words`?

File: rs/ethereum/cketh/minter/src/eth_logs/mod.rs [hunk 3]
```diff
@@ -424,3 +511,27 @@ fn parse_principal_from_slice(slice: &[u8]) -> Result<Principal, String> {
     }
     Principal::try_from_slice(principal_bytes).map_err(|err| err.to_string())
 }
+
+enum InternalLedgerSubaccountTag {}
+type InternalLedgerSubaccount = CheckedAmountOf<InternalLedgerSubaccountTag>;
+
+/// Ledger subaccount.
+///
+/// Internally represented as a u256 to optimize cbor encoding for low values,
```
Comment (THLO) [inline rs/ethereum/cketh/minter/src/eth_logs/mod.rs:520]: Again just for my understanding, why is a `u256` representation an optimization?

**Review by THLO — APPROVED** [review]

File: rs/ethereum/cketh/minter/src/eth_logs/mod.rs (hunk 1 above)
Comment (gregorydemay) [reply rs/ethereum/cketh/minter/src/eth_logs/mod.rs:465]: [`TryInto`](https://doc.rust-lang.org/std/convert/trait.TryInto.html) is a standard Rust trait, which is automatically implemented when the source type implements `TryFrom` (blanket implementation). So here, we want to convert a `Vec<T>` (`T` here is `[u8;32]`) into a fixed-size array `[T; N]`), which is obviously a fallible operation (hence it must implement `TryFrom` and not `From`) which is [implemented](https://doc.rust-lang.org/std/primitive.array.html#impl-TryFrom%3CVec%3CT,+A%3E%3E-for-%5BT;+N%5D) by the standard library. We know it shouldn't fail (hence the `unwrap`) because we ensure the correct length of the constructed vector.

File: rs/ethereum/cketh/minter/src/eth_logs/mod.rs (hunk 2 above)
Comment (gregorydemay) [reply rs/ethereum/cketh/minter/src/eth_logs/mod.rs:444]: I though one could alternatively use `32-byte words` (with hyphen) or `32 bytes words` (without hyphen). Are they not both correct?

File: rs/ethereum/cketh/minter/src/eth_logs/mod.rs (hunk 3 above)
Comment (gregorydemay) [reply rs/ethereum/cketh/minter/src/eth_logs/mod.rs:520]: It's because of how we do the cbor [serialization](https://github.com/dfinity/ic/blob/9fb40537484e9fd4d575e0b6d16cb164afab8b07/rs/ethereum/cketh/minter/src/cbor/u256.rs#L37).

File: rs/ethereum/cketh/minter/src/eth_logs/mod.rs (hunk 2 above)
Comment (ninegua) [reply rs/ethereum/cketh/minter/src/eth_logs/mod.rs:444]: Not related to the above comments, but this is a very cool use case of dependent types, namely, value depending on type. I didn't know Rust could do this!

**Review by ninegua — APPROVED** [review]

File: rs/ethereum/cketh/minter/src/eth_logs/mod.rs (hunk 2 above)
Comment (THLO) [reply rs/ethereum/cketh/minter/src/eth_logs/mod.rs:444]: I didn't manage to track down an authoritative explanation of the exact rule but a hyphen is always used when binding words that modify a noun (such as a "two-year-old child" where "two-year-old" functions as an adjective, see [here](https://ell.stackexchange.com/questions/43196/difference-between-two-years-old-and-two-year-old)). Similarly, nouns that modify nouns are also considered adjectives and therefore _usually_ appear in singular form, see [here](https://english.stackexchange.com/questions/258111/two-plural-nouns-in-a-row). An interesting exception is "arms race", where "arms" is plural to disambiguate.

In short, "32-byte words" is certainly the more commonly accepted form.
I claim that "32 bytes words" would sound unnatural to most native speakers.

Comment (gregorydemay) [reply rs/ethereum/cketh/minter/src/eth_logs/mod.rs:444]: Thanks a lot Thomas for this explanation! Let's remove the `s` then!
